packages/plugin-http-client/src/rest/AbstractResource.ts

```diff
@@ -0,0 +1,170 @@
+import { GenericError, Dependencies } from '@ima/core';
+
+declare module '@ima/core' {
+  interface OCAliasMap {
+    REST_CLIENT_BASE_API_URL: string;
+  }
+}
+
+import {
+  HttpClient,
+  HttpClientRequestMethod,
+  HttpClientRequestOptions,
+} from '../HttpClient';
+
+export class AbstractResource {
+  #httpClient: HttpClient;
+  #baseApiUrl: string;
+
+  static get $dependencies(): Dependencies {
+    return [HttpClient, '?REST_CLIENT_BASE_API_URL'];
+  }
+
+  static get PathType() {
+    return {
+      GET: 'get',
+      CREATE: 'create',
+      UPDATE: 'update',
+      REPLACE: 'replace',
+      DELETE: 'delete',
+    };
+  }
+
+  get path(): { [pathType: string]: string } {
+    throw new GenericError(
+      `AbstractResource: Static getter "path" must be overriden.`
+    );
+  }
+
+  get entityClass() {
+    return null;
+  }
+
+  constructor(httpClient: HttpClient, baseApiUrl: string) {
+    this.#httpClient = httpClient;
+
+    if (!baseApiUrl) {
+      throw new GenericError(`REST_CLIENT_BASE_API_URL is not set.`);
+    }
+    this.#baseApiUrl = baseApiUrl;
+  }
+
+  update(
+    data: object,
+    options: HttpClientRequestOptions,
+    pathType = (this.constructor as typeof AbstractResource).PathType.UPDATE
+  ) {
+    return this._restClientRequest('patch', pathType, data, options);
+  }
+
+  create(
+    data: object,
+    options: HttpClientRequestOptions,
+    pathType = (this.constructor as typeof AbstractResource).PathType.CREATE
+  ) {
+    return this._restClientRequest('post', pathType, data, options);
+  }
+
+  delete(
+    data: object,
+    options: HttpClientRequestOptions,
+    pathType = (this.constructor as typeof AbstractResource).PathType.DELETE
+  ) {
+    return this._restClientRequest('delete', pathType, data, options);
+  }
+
+  get(
+    data: object,
+    options: HttpClientRequestOptions,
+    pathType = (this.constructor as typeof AbstractResource).PathType.GET
+  ) {
+    return this._restClientRequest('get', pathType, data, options);
+  }
+
+  replace(
+    data: object,
+    options: HttpClientRequestOptions,
+    pathType = (this.constructor as typeof AbstractResource).PathType.REPLACE
+  ) {
+    return this._restClientRequest('put', pathType, data, options);
+  }
+
+  _getPathTemplate(pathType: string): string {
```

Comment:
Muze tu byt # misto _? P.S. I jinde asi kde je _, at to nemusim vypisovat :D

Reply:
anebo `private` kdyz jsme v tom typescripte xD

Reply:
Nahrazeno za #. Private je naprd, tím, že to používáme pak v JS, tak to pak nic neomezuje ne?